keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -413,28 +412,83 @@ On Pod Create/Update/Delete:
 
 ### Kube-Proxy
 
-Kube-proxy will be modified to consume EndpointSlice instances besides Endpoints resource. A flag will be added to kube-proxy to toggle the mode. 
-
-```
-Watch: Service, EndpointSlice ==> Manage: iptables, ipvs, etc
-```
-- Merge multiple EndpointSlice into an aggregated list.
-- Reuse the existing processing logic 
-
-### Endpoint Controller (classic)
-
-In order to ensure backward compatibility for external consumer of the core/v1 Endpoints API, the existing K8s endpoint controller will keep running until the API is EOL. The following limitations will apply:
-- Starting from EndpointSlice beta: If # of endpoints in one Endpoints object exceed 1000, generate a warning event to the object. 
-- Starting from EndpointSlice GA: Only include up to 1000 endpoints in one Endpoints Object and throw events.
+Kube-Proxy will support consuming EndpointSlice resources as an alternative to
+Endpoints resources. This will be enabled with the `EndpointSliceProxying`
+feature gate.
+
+This will require minimal changes to proxier implementations, instead updating
+the shared code that transforms Endpoints into a generic data structure to also
+transform EndpointSlices into that same data structure.
+
+### Endpoints Controller (classic)
+
+In order to ensure backwards compatibility for external consumer of the core/v1
+Endpoints API, the existing Endpoints controller will keep running until the
+API is EOL. As EndpointSlices become more stable, the Endpoints controller will
+gradually limit functionality.
+
+* Kubernetes 1.19: If the number of endpoints in one Endpoints object exceeds
+  1000, generated a warning event on the Endpoints resource.
+* Kubernetes 1.20: Limit the number of endpoints in one Endpoints object to
+  1000 and generate warning events on Endpoints resource if additional endpoints
+  are dropped.
+
+### EndpointsMirroring Controller
+
+In some cases, custom Endpoints resources are created by applications. To ensure
+that these applications will not need to immediately update their
+implementations to also create custome EndpointSlices, a new EndpointsMirroring
+controller will be used to mirror custome Endpoints resources to corresponding
+EndpointSlices.
+
+* A new `endpointsmirroring.kubernetes.io/skip` label will be introduced. When
+  this label is set to "true" on an Endpoints resource, the EndpointsMirroring
+  controller will not mirror this resource to an EndpointSlice.
+* The Endpoints controller will set this new label to "true" for all Endpoints
+  resources it manages.
```

Comment:
Describe what happens if an update is made to a service that causes the endpoints controller to being or stop managing updates to the associated endpoints resource.
* if the endpoints controller starts managing endpoints for a service and there is an existing endpoints object, will it add the label to it (I would expect so)
* if a service is modified (e.g. selector is removed?) so the endpoints controller is no longer managing the endpoints, does it remove the endpoints object, or remove the endpointslice label, or something else?

Reply:
Great questions. The first one is relatively straightforward, the Endpoints controller should automatically add that label to an Endpoints resource it is managing. I've added this in the testing plan.

The other question is harder to answer. This exposes a gap in the current behavior of both the [Endpoints](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L380) and [EndpointSlice](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go) controllers. Both will return early from `syncService` calls if a Service selector is nil. This is intentional for new Services, but also (likely unintentionally) means that removing a selector from a Service freezes corresponding Endpoints and EndpointSlices in the current state if/until a selector is set again on the Service. 

It would likely make sense to at least have the Endpoints controller remove a skip label on the corresponding Endpoints resource in this scenario, but it might actually have made more sense for this to actually result in the Endpoints and EndpointSlices getting deleted. I'm thinking that there is likely someone somewhere relying on this undocumented behavior though, so actually garbage collecting resources here is also dangerous. What do you think makes the most sense here?

Reply:
I've talked with @freehan about this and it sounds like the lack of garbage collection was likely a bug and not intentional. With that context, I think it makes sense to fix the bug in both the Endpoints and EndpointSlice controllers as part of the 1.19 release. I'll update this PR to reference that.

Reply:
Added to KEP and created tracking issue here: https://github.com/kubernetes/kubernetes/issues/91072.

Reply:
> Describe what happens if an update is made to a service that causes the endpoints controller to being or stop managing updates to the associated endpoints resource.
> 
> * if the endpoints controller starts managing endpoints for a service and there is an existing endpoints object, will it add the label to it (I would expect so)
> * if a service is modified (e.g. selector is removed?) so the endpoints controller is no longer managing the endpoints, does it remove the endpoints object, or remove the endpointslice label, or something else?

@liggitt is there any precedent on the API about this behaviour that we can follow? all the objects generated by a controller must be garbage collected once they don't belong to the parent object? 
I'm not familiar with the deployments/jobs controller, but may be we can find an idea on how they handle the pods,